nexus/reconfigurator/blippy/src/checks.rs

```diff
@@ -0,0 +1,524 @@
+            let kind = zone.zone_type.kind();
+            if let Some(previous) = transient_kinds_by_zpool
+                .entry(dataset.pool().id())
+                .or_default()
+                .insert(kind, zone)
+            {
+                blippy.push_note(
+                    Component::Sled(sled_id),
+                    Severity::Fatal,
+                    Kind::ZpoolFilesystemDatasetCollision {
+                        zone1: previous,
+                        zone2: zone,
+                        zpool: dataset.into_parts().0,
+                    },
+                );
+            }
+        }
+
+        // If a zone has both durable and transient datasets, they should be on
+        // the same pool.
+        match (zone.zone_type.durable_zpool(), zone.filesystem_pool.as_ref()) {
+            (Some(durable), Some(transient)) if durable != transient => {
+                blippy.push_note(
+                    Component::Sled(sled_id),
+                    Severity::Fatal,
+                    Kind::ZoneWithDatasetsOnDifferentZpools {
+                        zone,
+                        durable_zpool: durable.clone(),
+                        transient_zpool: transient.clone(),
+                    },
+                );
+            }
+            _ => (),
+        }
+    }
+}
+
+type DatasetByKind<'a> = BTreeMap<DatasetKind, &'a BlueprintDatasetConfig>;
+type DatasetsByZpool<'a> = BTreeMap<ZpoolUuid, DatasetByKind<'a>>;
+
+#[derive(Debug)]
+struct DatasetsBySled<'a> {
+    by_sled: BTreeMap<SledUuid, DatasetsByZpool<'a>>,
+    noted_sleds_missing_datasets: BTreeSet<SledUuid>,
+}
+
+impl<'a> DatasetsBySled<'a> {
+    fn new(blippy: &mut Blippy<'a>) -> Self {
+        let mut by_sled = BTreeMap::new();
+
+        for (&sled_id, config) in &blippy.blueprint().blueprint_datasets {
+            let by_zpool: &mut BTreeMap<_, _> =
+                by_sled.entry(sled_id).or_default();
+
+            for dataset in config.datasets.values() {
+                let by_kind: &mut BTreeMap<_, _> =
+                    by_zpool.entry(dataset.pool.id()).or_default();
+
+                match by_kind.entry(dataset.kind.clone()) {
+                    Entry::Vacant(slot) => {
+                        slot.insert(dataset);
+                    }
+                    Entry::Occupied(prev) => {
+                        blippy.push_note(
+                            Component::Sled(sled_id),
+                            Severity::Fatal,
+                            Kind::ZpoolWithDuplicateDatasetKinds {
+                                dataset1: prev.get(),
+                                dataset2: dataset,
+                                zpool: dataset.pool.id(),
+                            },
+                        );
+                    }
+                }
+            }
+        }
+
+        Self { by_sled, noted_sleds_missing_datasets: BTreeSet::new() }
+    }
+
+    fn for_sled(
```

Comment:
```suggestion
    fn check_sled_has_datasets(
```

(`for_sled` evokes a lookup, which this is, but I was a little surprised to see a check buried in here)